routes/api.js

```diff
@@ -84,24 +88,103 @@ var substituteCodeCell = function(payload) {
-};
+// var maskBytes = new Buffer(4);
+// var frameHeader = new Buffer(10);
+// var bufferList = new BufferList();
+// var wsconfig = {
+//     maxReceivedFrameSize: 0x100000  // 1MiB max frame size.
+// };
+
+// var wsconfig = {
+//     maxReceivedFrameSize: Math.pow(2, 64)
+// };
+
+function sendSocketData(conn, data) {
+    if (data.type === 'utf8') {
+        conn.sendUTF(data.utf8Data);
+    } else {
+        conn.sendBytes(data.binaryData);
+    }
+}
 
 function setupWSProxy(_server) {
     debug('setting up WebSocket proxy');
     server = _server;
 
+
+    var wsserver = new WebSocketServer({
+        httpServer: _server,
+        autoAcceptConnections: false,
+        maxReceivedFrameSize: Number.MAX_SAFE_INTEGER
+    });
+
+    wsserver.on('request', function(request) {
+        debug('ws connection request');
+
+        // XXX verify request.origin
+        var servConn = request.accept(null, request.origin);
+
+        var pendingServMsgs = [];
+        var bufferServMsgs = function(data) {
+            pendingServMsgs.push(data);
+        };
+        servConn.on('message', bufferServMsgs);
+        servConn.on('close', function(reasonCode, desc) {
+            debug('closing WS server connection:', reasonCode, desc);
+        });
+
+        var client = new WebSocketClient({
+            maxReceivedFrameSize: Number.MAX_SAFE_INTEGER
+        });
+
+        client.on('connect', function(clientConn) {
+            debug('ws client connected');
+
+            // INCOMING: kernel-gateway -> proxy -> client
+            clientConn.on('message', function(data) {
+                debug('INCOMING msg :: data length =', (data.utf8Data||data.binaryData).length);
+                sendSocketData(servConn, data);
+            });
+
+            // OUTGOING: client -> proxy -> kernel-gateway
+            servConn.on('message', function(data) {
+                debug('OUTGOING msg :: data length =', (data.utf8Data||data.binaryData).length);
+                processOutgoingMsg(data, function(newdata) {
+                    sendSocketData(clientConn, newdata);
+                });
+            });
+
+            // handle any buffered messages
+            servConn.removeListener('message', bufferServMsgs);
+            pendingServMsgs.forEach(function(data) {
+                sendSocketData(clientConn, data);
+            });
+        });
+
+        client.on('connectFailed', function(e) {
+            error('ws client failure', e);
+        });
+        client.on('error', function(e) {
+            error('ws client error', e);
+        });
+        client.on('close', function() {
+            debug('closing WS client connection');
+        });
+
+        var wsKgUrl = urljoin(kgUrl, kgBaseUrl, request.resourceURL.path).replace(/^http/, 'ws');
+        client.connect(wsKgUrl, null);
+    });
+
+    return;
+
+
     // Listen to the `upgrade` event and proxy the WebSocket requests as well.
     _server.on('upgrade', function(req, socket, head) {
```

Comment:
Yes, goes away. The `upgrade` is now handled internally by the `WebSocketServer/Client` classes.